crypto/asn1/a_time.c

```diff
@@ -385,62 +370,58 @@ int ASN1_TIME_set_string_X509(ASN1_TIME *s, const char *str)
 {
     ASN1_TIME t;
     struct tm tm;
-    int rv = 0;
     size_t len;
 
-    if ((len = strlen(str)) >= INT_MAX)
-        goto out;
-    t.length = (int)len;
-    t.data = (unsigned char *)str;
-    t.flags = ASN1_STRING_FLAG_X509_TIME;
-
-    t.type = V_ASN1_UTCTIME;
-
-    if (!ASN1_TIME_check(&t)) {
+    /* RFC 5280 4.1.2.5: Valid RFC5280 times must be either length 13 or 15. */
+    len = strlen(str);
+    switch (len) {
+    case 13:
+        t.type = V_ASN1_UTCTIME;
+        break;
+    case 15:
         t.type = V_ASN1_GENERALIZEDTIME;
-        if (!ASN1_TIME_check(&t))
-            goto out;
+        break;
+    default:
+        return 0;
     }
 
+    /* RFC 5280 4.1.2.5 Valid RFC5280 times must end in 'Z.' */
+#if defined(CHARSET_EBCDIC)
+    if (str[len - 1] != 0x5A)
+        return 0;
+#else
+    if (str[len - 1] != 'Z')
+        return 0;
+#endif
```

Comment:
This reads more easily to me:
```suggestion
    if (str[len - 1] != 0x5A)
        return 0;
```

Not need to comment that 5A is Z since it's in the comment above.